quinn-proto/src/connection/mod.rs

```diff
@@ -362,6 +362,20 @@ where
             return None;
         }
 
+        let mut num_datagrams = 0;
+        // TODO: Check whether this can be constant or there are situations where
+        // we have to enforce less packets due to path limitations. Seems like
+        // we are beyond anti-amplification check this might already be good.
+        //
+        // Also figure out how to make sure this is 1 on platforms which don't
+        // support GSO. Maybe it should be part of `ServerConfig`, and default
+        // to 1 on non-Linux. For Linux, should the kernel version and GSO
+        // support be detected or simply a minimum requirement?
+        // We can also keep setting the default on Linux to a reasonable number,
+        // and let applications overwrite if they run on an older kernel. Or
+        // let Quinn detect the Kernel version.
```

Comment:
Another option would be to take the maximum number of packets as an argument, though that could be relatively disruptive. That might be especially natural if we refactor to write packets into caller-supplied memory.

One way or another, it'd be nice to autodetect optimal behavior for most users. If UDP GSO support isn't near-universally deployed (isn't it pretty new?) we definitely shouldn't mandate it. There's also the possibility of batching outgoing packets unconditionally at this layer, and breaking it up later if the platform doesn't support it.

Reply:
I indeed thought about letting the caller pass a limit. But it seemed like it would just move the setting around.

> There's also the possibility of batching outgoing packets unconditionally at this layer, and breaking it up later if the platform doesn't support it.

Yes. The endpoint layer could then do multiple `sendmsg` calls by selecting the sub-slices. But it gets a bit more tricky if it gets `EAGAIN`  in the middle of such a transmission => Requires additional offset tracking. Or putting everything into a `Bytes` instance and splitting the `Transmit` again. But not producing it in the first place might be easiest.

Reply:
> it seemed like it would just move the setting around.

My thinking is that it would allow the parameter to be internal to the `quinn` layer, rather than exposed in the configuration object as a value that the user should generally not be manually adjusting.

> But it gets a bit more tricky if it gets EAGAIN in the middle of such a transmission

Ah, right, buffering is annoying, especially since it risks sending outdated information compared to composing only exactly the packets that will be sent.